Update sbt to 0.13.16; add Scala 2.12 support.

ucbjrl opened the pull request:
To clean, test, and build Scala 2.11 and Scala 2.12 versions:
  % sbt +clean +test +publishLocal

ucbjrl commented:
NOTE: You need to be running Java 8 to build the 2.12 version.

jackkoenig commented:
How do you build just the Scala 2.11 version? Also, if we use 2.12 features, will the 2.11 cross building still work?

ucbjrl commented:
The default is to build the 2.11 version.
```shell
% sbt clean test publishLocal
```
will only build the 2.11 version.

If the 2.12 features require source code changes which the 2.11 compiler doesn't understand, then cross building won't work.

jackkoenig commented:
Fantastic news, Scala 2.11.11 has a flag to let you use 2.12 features like the only one I really care about, implementing apply for case classes!!!

jackkoenig commented:
If the default is 2.11, how do you build 2.12? Also should we maybe make 2.12 the default since we want people to switch to it? Maybe make it the default on the official 1.0.0 release?

ucbjrl commented:
Check out [cross building](http://www.scala-sbt.org/0.13/docs/Cross-Build.html). It looks like:
```shell
  % sbt "++ 2.12.2" clean test
```
should do it.

We will publish both so there isn't a default as far as clients/users are concerned. The default build (plain sbt commands) is really for developers. Eventually 2.12 will be the default. Submit a PR to make it the default `scalaVersion := "2.12.3"` whenever you think we (developers) are ready.

jackkoenig commented on build.sbt:
```diff
@@ -20,19 +18,49 @@ version := "1.1-SNAPSHOT"
+    //  https://github.com/scala/bug/issues/10047
+    CrossVersion.partialVersion(scalaVersion) match {
+      case Some((2, scalaMajor: Int)) if scalaMajor < 12 => Seq()
+      case _ => Seq("-Xsource:2.11")
```
Is this necessary in Firrtl? It's necessary in chisel code using anonymous Bundles, but do the options have to be the same?

I ask because I actually want to add the "-Xsource:2.12" option when we compile with 2.11.11

ucbjrl commented:
It's not strictly necessary.

On Thu, Aug 24, 2017 at 2:06 PM, Jack Koenig <notifications@github.com>
wrote:

> *@jackkoenig* commented on this pull request.
> ------------------------------
>
> In build.sbt
> <https://github.com/freechipsproject/firrtl/pull/639#discussion_r135130964>
> :
>
> > @@ -20,19 +18,49 @@ version := "1.1-SNAPSHOT"
>
>  scalaVersion := "2.11.11"
>
> -javacOptions ++= Seq("-source", "1.7", "-target", "1.7")
> +crossScalaVersions := Seq("2.11.11", "2.12.3")
> +
> +def scalacOptionsVersion(scalaVersion: String): Seq[String] = {
> +  Seq() ++ {
> +    // If we're building with Scala > 2.11, enable the compile option
> +    //  switch to support our anonymous Bundle definitions:
> +    //  https://github.com/scala/bug/issues/10047
> +    CrossVersion.partialVersion(scalaVersion) match {
> +      case Some((2, scalaMajor: Int)) if scalaMajor < 12 => Seq()
> +      case _ => Seq("-Xsource:2.11")
>
> Is this necessary in Firrtl? It's necessary in chisel code using anonymous
> Bundles, but do the options have to be the same?
>
> I ask because I actually want to add the "-Xsource:2.12" option when we
> compile with 2.11.11
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/freechipsproject/firrtl/pull/639#pullrequestreview-58510637>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGh1wLTF3z83_OE0qoKw_mSpb1N-DxgMks5sbeXmgaJpZM4PAVgr>
> .
>

ucbjrl commented:
This is failing because Travis appears to be waiting to run the "prepare cache-verilator" stage until the test stages complete, but they require verilator. I suspect this is due to the cross versioning, even though the travis.yml isn't requesting a cross version build.

Could a Travis expert take a look at this please?

ucbjrl commented:
Could someone approve this now that it's passing Travis? I'd like to get it in next weeks release.

jackkoenig commented:
Is there an equivalent rocket-chip branch that uses this firrtl branch as well as the related chisel one?

ucbjrl commented:
Not yet, but do you want to hold off a Scala 2.12 release until we've
successfully build rocketchip with it?


On Mon, Sep 11, 2017 at 11:26 AM, Jack Koenig <notifications@github.com>
wrote:

> Is there an equivalent rocket-chip branch that uses this firrtl branch as
> well as the related chisel one?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/freechipsproject/firrtl/pull/639#issuecomment-328616323>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGh1wC_B1KiqEyMPc5bOSJNYS7mP2yv2ks5shXtbgaJpZM4PAVgr>
> .
>

jackkoenig commented:
Fair point. I don't think we should hold off but we should verify that rocket-chip works with these branches just to be sure we're cross-building correctly

The branch of rocket-chip that has bumped chisel3 and firrtl is: https://github.com/freechipsproject/rocket-chip/tree/bump-chisel-firrtl

jackkoenig commented:
Using the above rocket-chip branch with this branch of chisel3 merged into master (and the equivalent in firrtl), I get the following error:

```
/scratch/jack/rocket-chip-bump/chisel3/build.sbt:128: error: not found: value ScalaUnidocPlugin
  enablePlugins(ScalaUnidocPlugin).
```
Seems like the way plugins get enabled changed? Probably a fairly simple update of sbt configuration in rocket-chip should do it?

ucbjrl commented:
Updated plugins have a different interface. I've created
a Scala212-almostcompatible branch in rocket-chip with the Scala212
versions of chisel3 and firrtl, updated plugins, use of Scala 2.11.11 and
sbt 0.13.16 and it builds and runs the emulation tests (so far).

It's "almostcompatible" since it will require changes for an updated
sbt-pack plugin, and eventual replacement of the project/build.scala with
an appropriate build.sbt file (for sbt 1.x).


On Mon, Sep 11, 2017 at 6:03 PM, Jack Koenig <notifications@github.com>
wrote:

> Using the above rocket-chip branch with this branch of chisel3 merged into
> master (and the equivalent in firrtl), I get the following error:
>
> /scratch/jack/rocket-chip-bump/chisel3/build.sbt:128: error: not found: value ScalaUnidocPlugin
>   enablePlugins(ScalaUnidocPlugin).
>
> Seems like the way plugins get enabled changed? Probably a fairly simple
> update of sbt configuration in rocket-chip should do it?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/freechipsproject/firrtl/pull/639#issuecomment-328700707>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGh1wNirpIZ2asT6ZO6caox7GvDhi4ouks5shdhMgaJpZM4PAVgr>
> .
>
